typewriter/fixes/fixer_utils.py

```diff
@@ -133,3 +145,221 @@ def touch_import(package, name, node):
+        self.node = node
+
+    def __hash__(self):
+        return hash((self.node.get_lineno(),
+                     self.node.depth(),
+                     self.node.__str__()))
+
+
+@cache
+def _get_import_info_cachable(hashable_node):
+    # type: (HashableNode) -> ImportNodeInfo
+    """
+    We cache using hashable_node as a key and analyze hashable_node.node
+
+    If node is a valid import_stmt, this will return a named tuple for each component of the
+    statement
+        ex: from a.b import c as d, e  => (imports={'a.b': [(import_name='c',
+                                                             binding_name='d'),
+                                                            (import_name='e',
+                                                             binding_name='e')]},
+                                           node=node)
+
+    Since we will regularly iterate over the list import node for their information when
+    searching for binding or import matches, we cache the results of this function for
+    quick and non-redundant lookups
+
+    Parameters
+    -----------
+    hashable_node : HashableNode
+
+    Return
+    -----------
+    ImportNodeInfo
+    """
+    # Get the node from the hashable_node wrapper
+    node = hashable_node.node
+
+    def handle_name(node):
+        if node.type in (syms.dotted_as_name, syms.import_as_name):
+            # [Grammar] dotted_as_name: dotted_name ['as' NAME]
+            package = import_name = binding_name = None
+            name = node.children[0]
+            binding_name = str(node.children[2]).strip()
+            if name.type in (syms.dotted_name, token.NAME):
+                # [Grammar] dotted_name: NAME ('.' NAME)*
+                # We don't need the third field since we know the alias will be the binding_name
+                package, import_name, _ = decompose_name(name)
+            return package, import_name, binding_name
+        # As long as we aren't ignoring dotted names
+        # If there was no "as", we know this is dotted_name
+        # [Grammar] dotted_name: NAME ('.' NAME)*
+        return decompose_name(node)
+
+    package = None
+    imports = dict()  # type: Dict[str, Set[ImportPairing]]
+    if node.type == syms.import_name:
+        # [Grammar]: import_name: 'import' dotted_as_names
+        as_name = node.children[1]
+        if as_name.type in (syms.dotted_as_names, syms.import_as_names):
+            # [Grammar]: dotted_as_names: dotted_as_name (',' dotted_as_name)*
+            for child in as_name.children:
+                if str(child).strip() == ',':
+                    continue
+                package, import_name, binding_name = handle_name(child)
+                if package is None:
+                    package = 'no_pkg'
+                imports[package] = imports.get(package, set())
+                imports[package].add(ImportPairing(import_name, binding_name))
+        package, import_name, binding_name = handle_name(as_name)
```

Comment:
So, I think this whole chunk here, after the `if as_name.type in..` block above, should be in it's own `else` block.

However... seeing how similar the code in these two chunks is, I think we can unify the paths, by just doing something like:

```
if as_name.type in (syms.dotted_as_names, syms.import_as_names):
    # [Grammar]: dotted_as_names: dotted_as_name (',' dotted_as_name)*
    as_names = [child for child in as_name.children if str(child).strip() != ',']
else:
    as_names = [as_name]
```
...and then doing the for loop over as_names

Reply:
nice, very clean